File: backend/app/schemas/tally.py
"""Module 12 (Data Migration) — Tally import schemas."""

import uuid
from datetime import date, datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.common import DocumentMeta, ORMModel


# --- Coverage catalogue --------------------------------------------------------------


class TallyEntityCoverage(BaseModel):
    """One row of the Tally -> OptiERP coverage matrix."""

    key: str
    label: str
    tally_tag: str
    target: str
    module: str
    stage: int
    support: str  # full | partial | reference | none
    notes: str


class TallyCatalogueResponse(BaseModel):
    entities: list[TallyEntityCoverage]
    #: Tally reserved group -> {root_type, account_type, party_type}
    primary_groups: dict[str, dict[str, str | None]]
    #: Tally voucher type -> target DocType
    voucher_types: dict[str, str]
    modules: list[str]


# --- Sessions ------------------------------------------------------------------------


class TallyImportOptions(BaseModel):
    """Run options stored on the session."""

    submit_vouchers: bool = True  # False leaves everything as drafts
    default_warehouse_id: uuid.UUID | None = None
    order_lead_days: int = Field(default=7, ge=0, le=365)


class TallyImportCreate(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    opening_date: date | None = None
    options: TallyImportOptions = Field(default_factory=TallyImportOptions)


class TallyImportEntityResponse(DocumentMeta):
    entity_key: str
    label: str
    target_doctype: str
    module: str
    stage: int
    support: str
    selected: bool
    total: int
    created: int
    updated: int
    skipped: int
    failed: int


class TallyImportListItem(ORMModel):
    id: uuid.UUID
    name: str
    title: str | None
    source_type: str
    file_name: str | None
    tally_company_name: str | None
    from_date: date | None
    to_date: date | None
    status: str
    total_records: int
    imported_count: int
    skipped_count: int
    error_count: int
    creation: datetime
    finished_at: datetime | None


class TallyImportResponse(DocumentMeta):
    name: str
    title: str | None
    company_id: uuid.UUID
    source_type: str
    file_name: str | None
    file_size: int
    tally_company_name: str | None
    from_date: date | None
    to_date: date | None
    opening_date: date | None
    status: str
    options: dict[str, Any] | None
    total_records: int
    imported_count: int
    skipped_count: int
    error_count: int
    started_at: datetime | None
    finished_at: datetime | None
    error_message: str | None
    entities: list[TallyImportEntityResponse] = Field(default_factory=list)


class TallyImportLogResponse(ORMModel):
    id: uuid.UUID
    creation: datetime
    phase: str
    entity_key: str | None
    level: str
    message: str
    context: dict[str, Any] | None


class TallyImportSummaryResponse(BaseModel):
    session: TallyImportResponse
    entities: list[TallyImportEntityResponse]
    logs: list[TallyImportLogResponse]
    status_counts: dict[str, int]
    unmapped_count: int
    low_confidence_count: int


class TallyEntitySelection(BaseModel):
    """Turn entities on/off before a run: {"voucher_sales": true, ...}"""

    selections: dict[str, bool]


# --- Staging -------------------------------------------------------------------------


class TallyStagingMessage(BaseModel):
    level: str
    message: str
    field: str | None = None


class TallyStagingRecordResponse(ORMModel):
    id: uuid.UUID
    entity_key: str
    sequence: int
    tally_guid: str | None
    tally_name: str | None
    tally_parent: str | None
    tally_voucher_type: str | None
    voucher_number: str | None
    posting_date: date | None
    amount: Decimal | None
    target_doctype: str | None
    target_id: uuid.UUID | None
    target_name: str | None
    status: str
    messages: list[dict[str, Any]] | None


class TallyStagingListResponse(BaseModel):
    items: list[TallyStagingRecordResponse]
    total: int
    limit: int
    offset: int


class TallyStagingDetailResponse(TallyStagingRecordResponse):
    """A single staging row including the raw Tally payload, for debugging."""

    raw: dict[str, Any] | None
    normalised: dict[str, Any] | None


# --- Mappings ------------------------------------------------------------------------


class TallyMappingResponse(DocumentMeta):
    entity_key: str
    tally_name: str
    tally_guid: str | None
    tally_parent: str | None
    target_doctype: str
    target_id: uuid.UUID | None
    target_name: str | None
    match_method: str
    confidence: int
    is_locked: bool
    attributes: dict[str, Any] | None
    notes: str | None


class TallyMappingUpdate(BaseModel):
    """A tester's explicit mapping decision (locks the row)."""

    target_doctype: str
    target_id: uuid.UUID | None = None
    notes: str | None = None


class TallyMappingCreate(TallyMappingUpdate):
    entity_key: str
    tally_name: str = Field(min_length=1, max_length=255)


class TallyMappingBulkUpdate(BaseModel):
    items: list[dict[str, Any]] = Field(min_length=1)


# --- Phase results -------------------------------------------------------------------


class TallyAutoMapResponse(BaseModel):
    summary: dict[str, dict[str, int]]
    session: TallyImportResponse


class TallyValidateResponse(BaseModel):
    planned: dict[str, int]
    total_planned: int
    unresolved: dict[str, list[str]]
    unresolved_count: int
    blockers: list[str]
    session: TallyImportResponse


class TallyRollbackResponse(BaseModel):
    cancelled: int
    failed: list[str]
    session: TallyImportResponse


class TallyWorkspaceStats(BaseModel):
    """Numbers for the Data Migration workspace card."""

    total_imports: int
    completed_imports: int
    failed_imports: int
    documents_imported: int
    unmapped_names: int
    last_import_at: datetime | None
    last_import_status: str | None
    supported_entities: int

File: frontend/src/types/tally.ts
// Module 12 — Data Migration (Tally import) types. Mirrors app/schemas/tally.py.

export type TallySupport = "full" | "partial" | "reference" | "none";

export type TallyImportStatus =
  | "Draft"
  | "Parsed"
  | "Mapped"
  | "Validated"
  | "Importing"
  | "Imported"
  | "Partially Imported"
  | "Failed"
  | "Rolled Back";

export type TallyRecordStatus =
  | "Pending"
  | "Ready"
  | "Warning"
  | "Error"
  | "Imported"
  | "Skipped"
  | "Rolled Back";

export interface TallyEntityCoverage {
  key: string;
  label: string;
  tally_tag: string;
  target: string;
  module: string;
  stage: number;
  support: TallySupport;
  notes: string;
}

export interface TallyCatalogue {
  entities: TallyEntityCoverage[];
  primary_groups: Record<
    string,
    { root_type: string; account_type: string | null; party_type: string | null }
  >;
  voucher_types: Record<string, string>;
  modules: string[];
}

export interface TallyImportEntity {
  id: string;
  entity_key: string;
  label: string;
  target_doctype: string;
  module: string;
  stage: number;
  support: TallySupport;
  selected: boolean;
  total: number;
  created: number;
  updated: number;
  skipped: number;
  failed: number;
}

export interface TallyImportListItem {
  id: string;
  name: string;
  title: string | null;
  source_type: string;
  file_name: string | null;
  tally_company_name: string | null;
  from_date: string | null;
  to_date: string | null;
  status: TallyImportStatus;
  total_records: number;
  imported_count: number;
  skipped_count: number;
  error_count: number;
  creation: string;
  finished_at: string | null;
}

export interface TallyImport extends TallyImportListItem {
  company_id: string;
  file_size: number;
  opening_date: string | null;
  options: Record<string, unknown> | null;
  started_at: string | null;
  error_message: string | null;
  entities: TallyImportEntity[];
}

export interface TallyImportLog {
  id: string;
  creation: string;
  phase: string;
  entity_key: string | null;
  level: "info" | "warning" | "error";
  message: string;
  context: Record<string, unknown> | null;
}

export interface TallyImportSummary {
  session: TallyImport;
  entities: TallyImportEntity[];
  logs: TallyImportLog[];
  status_counts: Record<string, number>;
  unmapped_count: number;
  low_confidence_count: number;
}

export interface TallyStagingMessage {
  level: "info" | "warning" | "error";
  message: string;
  field: string | null;
}

export interface TallyStagingRecord {
  id: string;
  entity_key: string;
  sequence: number;
  tally_guid: string | null;
  tally_name: string | null;
  tally_parent: string | null;
  tally_voucher_type: string | null;
  voucher_number: string | null;
  posting_date: string | null;
  amount: string | null;
  target_doctype: string | null;
  target_id: string | null;
  target_name: string | null;
  status: TallyRecordStatus;
  messages: TallyStagingMessage[] | null;
}

export interface TallyStagingList {
  items: TallyStagingRecord[];
  total: number;
  limit: number;
  offset: number;
}

export interface TallyMapping {
  id: string;
  entity_key: string;
  tally_name: string;
  tally_guid: string | null;
  tally_parent: string | null;
  target_doctype: string;
  target_id: string | null;
  target_name: string | null;
  match_method: "guid" | "exact" | "normalised" | "fuzzy" | "created" | "manual" | "auto";
  confidence: number;
  is_locked: boolean;
  attributes: Record<string, unknown> | null;
  notes: string | null;
}

export interface TallyMappingTarget {
  id: string;
  name: string;
  [key: string]: string;
}

export interface TallyAutoMapResult {
  summary: Record<
    string,
    { matched: number; to_create: number; locked: number; low_confidence: number }
  >;
  session: TallyImport;
}

export interface TallyValidateResult {
  planned: Record<string, number>;
  total_planned: number;
  unresolved: Record<string, string[]>;
  unresolved_count: number;
  blockers: string[];
  session: TallyImport;
}

export interface TallyRollbackResult {
  cancelled: number;
  failed: string[];
  session: TallyImport;
}

/** Shape every module workspace endpoint returns (see ModuleWorkspace.vue). */
export interface WorkspaceStats {
  currency: string;
  chart_title: string;
  trend_format?: "int" | "currency";
  cards: { label: string; value: number; format: "int" | "currency" }[];
  trend: { label: string; value: number }[];
}

/**
 * Metadata-engine slug (`/m/<slug>`) -> the Tally entity that fills it.
 * Drives the "Import from Tally" button on every generic master list, so a new
 * master screen gets the option without touching its view.
 */
export const SLUG_TO_TALLY_ENTITY: Record<string, { entity: string; module: string }> = {
  customer: { entity: "customer", module: "selling" },
  supplier: { entity: "supplier", module: "buying" },
  "cost-center": { entity: "cost_centre", module: "accounting" },
  "item-group": { entity: "stock_group", module: "stock" },
  address: { entity: "customer", module: "selling" },
  contact: { entity: "customer", module: "selling" },
};

/** Which catalogue entities each module's "Import from Tally" button pre-selects. */
export const MODULE_ENTITY_HINTS: Record<string, string[]> = {
  accounting: [
    "group",
    "ledger",
    "cost_centre",
    "opening_ledger",
    "voucher_sales",
    "voucher_purchase",
    "voucher_credit_note",
    "voucher_debit_note",
    "voucher_journal",
    "voucher_receipt",
    "voucher_payment",
    "voucher_contra",
  ],
  selling: ["customer", "voucher_sales", "voucher_credit_note", "voucher_sales_order"],
  buying: ["supplier", "voucher_purchase", "voucher_debit_note", "voucher_purchase_order"],
  stock: [
    "unit",
    "stock_group",
    "godown",
    "stock_item",
    "price_list",
    "opening_stock",
    "voucher_delivery_note",
    "voucher_receipt_note",
    "voucher_stock_journal",
    "voucher_physical_stock",
  ],
  manufacturing: ["stock_item", "voucher_manufacturing_journal", "voucher_stock_journal"],
};
